modules/ingest-common/src/main/java/org/opensearch/ingest/common/FingerprintProcessor.java

```diff
@@ -0,0 +1,279 @@
+ */
+public final class FingerprintProcessor extends AbstractProcessor {
+    public static final String TYPE = "fingerprint";
+    // this processor is introduced in 2.16.0, we append the OpenSearch version to the hash method name to ensure
+    // that this processor always generates same hash value based on a specific hash method, if the processing logic
+    // of this processor changes in future version, the version number in the hash method should be increased correspondingly.
+    private static final Set<String> HASH_METHODS = Set.of("MD5@2.16.0", "SHA-1@2.16.0", "SHA-256@2.16.0", "SHA3-256@2.16.0");
+
+    // fields used to generate hash value
+    private final List<String> fields;
+    // all fields other than the excluded fields are used to generate hash value
+    private final List<String> excludeFields;
+    // the target field to store the hash value, defaults to fingerprint
+    private final String targetField;
+    // hash method used to generate the hash value, defaults to SHA-1
+    private final String hashMethod;
+    private final boolean ignoreMissing;
+
+    FingerprintProcessor(
+        String tag,
+        String description,
+        @Nullable List<String> fields,
+        @Nullable List<String> excludeFields,
+        String targetField,
+        String hashMethod,
+        boolean ignoreMissing
+    ) {
+        super(tag, description);
+        if (fields != null && !fields.isEmpty()) {
+            if (fields.stream().anyMatch(Strings::isNullOrEmpty)) {
+                throw new IllegalArgumentException("field name in [fields] cannot be null nor empty");
+            }
+            if (excludeFields != null && !excludeFields.isEmpty()) {
+                throw new IllegalArgumentException("either fields or exclude_fields can be set");
+            }
+        }
+        if (excludeFields != null && !excludeFields.isEmpty() && excludeFields.stream().anyMatch(Strings::isNullOrEmpty)) {
+            throw new IllegalArgumentException("field name in [exclude_fields] cannot be null nor empty");
+        }
+
+        if (!HASH_METHODS.contains(hashMethod.toUpperCase(Locale.ROOT))) {
+            throw new IllegalArgumentException("hash method must be MD5@2.16.0, SHA-1@2.16.0 or SHA-256@2.16.0 or SHA3-256@2.16.0");
+        }
+        this.fields = fields;
+        this.excludeFields = excludeFields;
+        this.targetField = targetField;
+        this.hashMethod = hashMethod;
+        this.ignoreMissing = ignoreMissing;
+    }
+
+    public List<String> getFields() {
+        return fields;
+    }
+
+    public List<String> getExcludeFields() {
+        return excludeFields;
+    }
+
+    public String getTargetField() {
+        return targetField;
+    }
+
+    public String getHashMethod() {
+        return hashMethod;
+    }
+
+    public boolean isIgnoreMissing() {
+        return ignoreMissing;
+    }
+
+    @Override
+    public IngestDocument execute(IngestDocument document) {
```

Comment:
Thanks @gaobinlong, definitely good progress here!

The risk with this type of code is that a refactoring can subtly change the output and we would want tests to catch this. The general pattern I've seen is that you have a test matrix that has all relevant dimensions: input document (this needs to have a variety of input that hits all the code paths below: fields, exclude fields, nested documents, etc) and hash method. So in this case there might be 5 or 10 permutations of documents and 4 options for hash method. For each one of these cases, there should be a hardcoded fingerprint value in the tests. We expect that fingerprint to be stable over time, so if code change results in the test failing then we know that's a problem.

You did a bit of this in the YAML test by testing the fingerprint output, but I think we need to be comprehensive here so that we can detect if a future refactoring inadvertently makes a breaking change. Does this make sense?

Reply:
Thanks, I've added more yaml test cases to verify the hash value, please help to take a look!